# 01. Importing libraries

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

# 02. Data types

In [2]:
path = r'C:\Users\Kozole\Desktop\CF Kurs\Excercise\4\Instacart Basket Analysis'

In [3]:
ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged_4.7.pkl'))

In [4]:
ords_prods_merge.shape

(32406041, 18)

# 03. Task 4.8

In [7]:
# 02. In this Exercise, you learned how to find the aggregated mean of the “order_number” 
# column grouped by “department_id” for a subset of your dataframe. Now, repeat this process for the entire dataframe.

In [5]:
# Grouping data by “department_id” and finding the aggregated mean

ords_prods_merge.groupby('department_id')['order_number'].mean()

department_id
1     15.457838
2     17.277920
3     17.170395
4     17.811403
5     15.213779
6     16.439806
7     17.225802
8     15.340650
9     15.895474
10    20.197148
11    16.171095
12    15.887671
13    16.583536
14    16.771241
15    16.165037
16    17.665606
17    15.694469
18    19.310397
19    17.177343
20    16.473447
21    22.902379
Name: order_number, dtype: float64

In [9]:
# 03. Analyze the result. How do the results for the entire dataframe differ from those of the subset? 
# Include your comments in a markdown cell below the executed code.

# The values of the means of the entire dataframe and the subset are not that different and all are in range of +1 or -1 except for department_id=21 that had value of 25.5 in the subset and 22.9 for the entire dataframe.

In [10]:
# Follow the instructions in the Exercise for creating a loyalty flag for existing 
# customers using the transform() and loc() functions.

In [6]:
# Grouping the dataframe by user_id and order_number and creating new column max_order

ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

In [7]:
ords_prods_merge.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   3139998      138            28                   6                 11   
1   1977647      138            30                   6                 17   
2    389851      709             2                   0                 21   
3    652770      764             1                   3                 13   
4   1813452      764             3                   4                 17   

   days_since_prior_order  add_to_cart_order  reordered _merge  \
0                     3.0                  5          0   both   
1                    20.0                  1          1   both   
2                     6.0                 20          0   both   
3                     NaN                 10          0   both   
4                     9.0                 11          1   both   

     price_range_loc     busiest_day    busiest_days busiest_period_of_day  \
0  Mid-range product  Regularly busy  Regularly busy        Average orders   
1  Mid-range product  Regularly busy  Regularly busy        Average orders   
2  Mid-range product     Busiest day    Busiest days        Average orders   
3  Mid-range product  Regularly busy    Slowest days        Average orders   
4  Mid-range product      Least busy    Slowest days        Average orders   

   max_order  
0         32  
1         32  
2          5  
3          3  
4          3

In [8]:
# Adding loyalty_flag Loyal customer

ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [9]:
# Adding loyalty_flag Regular customer

ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

In [10]:
# Adding loyalty_flag New customer

ords_prods_merge.loc[ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [11]:
# Checking if right flags ware assigned

ords_prods_merge[['user_id', 'loyalty_flag', 'order_number']].head(15)

user_id      loyalty_flag  order_number
0       138  Regular customer            28
1       138  Regular customer            30
2       709      New customer             2
3       764      New customer             1
4       764      New customer             3
5       777  Regular customer            16
6       825      New customer             3
7       910  Regular customer            12
8      1052  Regular customer            10
9      1052  Regular customer            15
10     1379      New customer             1
11     1480    Loyal customer            36
12     1480    Loyal customer            38
13     1480    Loyal customer            39
14     1494  Regular customer             6

In [20]:
# 05. The marketing team at Instacart wants to know whether there’s a difference between the 
# spending habits of the three types of customers you identified. 
# Use the loyalty flag you created and check the basic statistics of the product prices for each 
# loyalty category (Loyal Customer, Regular Customer, and New Customer). 
# What you’re trying to determine is whether the prices of products purchased by loyal customers differ 
# from those purchased by regular or new customers.

In [12]:
# Checking the basic statistics of the product prices for each loyalty category

ords_prods_merge.groupby('loyalty_flag').agg({'prices': ['mean', 'min', 'max']})

prices              
                       mean  min      max
loyalty_flag                             
Loyal customer    10.386229  1.0  99999.0
New customer      13.294406  1.0  99999.0
Regular customer  12.495501  1.0  99999.0

# The average price of products purchased by Loyal Customers is lower than the products purchased by Regular or New Customers

In [22]:
# 06. The team now wants to target different types of spenders in their marketing campaigns. 
# This can be achieved by looking at the prices of the items people are buying. 
# Create a spending flag for each user based on the average price across all their orders using the following criteria:

# If the mean of the prices of products purchased by a user is lower than 10, then flag them as a “Low spender.”
# If the mean of the prices of products purchased by a user is higher than or equal to 10, then flag them as a “High spender.”

In [13]:
# Grouping the dataframe by user_id and prices and creating new column average_price

ords_prods_merge['average_price'] = ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)

In [14]:
ords_prods_merge.head(15)

product_id                product_name  aisle_id  department_id  prices  \
0            1  Chocolate Sandwich Cookies        61             19     5.8   
1            1  Chocolate Sandwich Cookies        61             19     5.8   
2            1  Chocolate Sandwich Cookies        61             19     5.8   
3            1  Chocolate Sandwich Cookies        61             19     5.8   
4            1  Chocolate Sandwich Cookies        61             19     5.8   
5            1  Chocolate Sandwich Cookies        61             19     5.8   
6            1  Chocolate Sandwich Cookies        61             19     5.8   
7            1  Chocolate Sandwich Cookies        61             19     5.8   
8            1  Chocolate Sandwich Cookies        61             19     5.8   
9            1  Chocolate Sandwich Cookies        61             19     5.8   
10           1  Chocolate Sandwich Cookies        61             19     5.8   
11           1  Chocolate Sandwich Cookies        61             19     5.8   
12           1  Chocolate Sandwich Cookies        61             19     5.8   
13           1  Chocolate Sandwich Cookies        61             19     5.8   
14           1  Chocolate Sandwich Cookies        61             19     5.8   

    order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    3139998      138            28                   6                 11   
1    1977647      138            30                   6                 17   
2     389851      709             2                   0                 21   
3     652770      764             1                   3                 13   
4    1813452      764             3                   4                 17   
5    1701441      777            16                   1                  7   
6    1871483      825             3                   2                 14   
7    1290456      910            12                   3                 10   
8     369558     1052            10                   1                 20   
9     589712     1052            15                   1                 12   
10   1798115     1379             1                   5                 11   
11    497193     1480            36                   3                 13   
12   1727826     1480            38                   4                 13   
13   2694698     1480            39                   2                 12   
14   3171753     1494             6                   3                 10   

    ...  add_to_cart_order  reordered  _merge    price_range_loc  \
0   ...                  5          0    both  Mid-range product   
1   ...                  1          1    both  Mid-range product   
2   ...                 20          0    both  Mid-range product   
3   ...                 10          0    both  Mid-range product   
4   ...                 11          1    both  Mid-range product   
5   ...                  7          0    both  Mid-range product   
6   ...                  2          0    both  Mid-range product   
7   ...                  1          0    both  Mid-range product   
8   ...                  1          0    both  Mid-range product   
9   ...                  2          1    both  Mid-range product   
10  ...                  1          0    both  Mid-range product   
11  ...                  1          0    both  Mid-range product   
12  ...                  3          1    both  Mid-range product   
13  ...                  4          1    both  Mid-range product   
14  ...                  8          0    both  Mid-range product   

       busiest_day    busiest_days busiest_period_of_day max_order  \
0   Regularly busy  Regularly busy        Average orders        32   
1   Regularly busy  Regularly busy        Average orders        32   
2      Busiest day    Busiest days        Average orders         5   
3   Regularly busy    Slowest days        Average orders         3   
4       Least busy    Slowest days        Average o

In [15]:
# Adding spenders_flag Low spender

ords_prods_merge.loc[ords_prods_merge['average_price'] < 10, 'spenders_flag'] = 'Low spender'

In [16]:
# Adding spenders_flag High spender

ords_prods_merge.loc[ords_prods_merge['average_price'] >= 10, 'spenders_flag'] = 'High spender'

In [17]:
ords_prods_merge['spenders_flag'].value_counts(dropna=False)

Low spender     31771789
High spender      634252
Name: spenders_flag, dtype: int64

In [18]:
ords_prods_merge.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   3139998      138            28                   6                 11   
1   1977647      138            30                   6                 17   
2    389851      709             2                   0                 21   
3    652770      764             1                   3                 13   
4   1813452      764             3                   4                 17   

   ...  reordered  _merge    price_range_loc     busiest_day    busiest_days  \
0  ...          0    both  Mid-range product  Regularly busy  Regularly busy   
1  ...          1    both  Mid-range product  Regularly busy  Regularly busy   
2  ...          0    both  Mid-range product     Busiest day    Busiest days   
3  ...          0    both  Mid-range product  Regularly busy    Slowest days   
4  ...          1    both  Mid-range product      Least busy    Slowest days   

  busiest_period_of_day max_order      loyalty_flag  average_price  \
0        Average orders        32  Regular customer       6.935811   
1        Average orders        32  Regular customer       6.935811   
2        Average orders         5      New customer       7.930208   
3        Average orders         3      New customer       4.972414   
4        Average orders         3      New customer       4.972414   

  spenders_flag  
0   Low spender  
1   Low spender  
2   Low spender  
3   Low spender  
4   Low spender  

[5 rows x 22 columns]

In [23]:
# 07.In order to send relevant notifications to users within the app 
# (for instance, asking users if they want to buy the same item again), the Instacart team wants you to determine frequent 
# versus non-frequent customers. Create an order frequency flag that marks the regularity of a user’s ordering behavior 
# according to the median in the “days_since_prior_order” column. The criteria for the flag should be as follows:

# If the median of “days_since_prior_order” is higher than 20, then the customer should be labeled a “Non-frequent customer.”
# If the median is higher than 10 and lower than or equal to 20, then the customer should be labeled a “Regular customer.”
# If the median is lower than or equal to 10, then the customer should be labeled a “Frequent customer.”

In [19]:
# Grouping the dataframe by user_id and days_since_prior_order and creating new column frequency_median

ords_prods_merge['frequency_median'] = ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)

In [20]:
ords_prods_merge.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   3139998      138            28                   6                 11   
1   1977647      138            30                   6                 17   
2    389851      709             2                   0                 21   
3    652770      764             1                   3                 13   
4   1813452      764             3                   4                 17   

   ...  _merge    price_range_loc     busiest_day    busiest_days  \
0  ...    both  Mid-range product  Regularly busy  Regularly busy   
1  ...    both  Mid-range product  Regularly busy  Regularly busy   
2  ...    both  Mid-range product     Busiest day    Busiest days   
3  ...    both  Mid-range product  Regularly busy    Slowest days   
4  ...    both  Mid-range product      Least busy    Slowest days   

  busiest_period_of_day max_order      loyalty_flag average_price  \
0        Average orders        32  Regular customer      6.935811   
1        Average orders        32  Regular customer      6.935811   
2        Average orders         5      New customer      7.930208   
3        Average orders         3      New customer      4.972414   
4        Average orders         3      New customer      4.972414   

   spenders_flag frequency_median  
0    Low spender              8.0  
1    Low spender              8.0  
2    Low spender              8.0  
3    Low spender              9.0  
4    Low spender              9.0  

[5 rows x 23 columns]

In [21]:
# Adding frequency_flag Non-frequent customer

ords_prods_merge.loc[ords_prods_merge['frequency_median'] > 20, 'frequency_flag'] = 'Non-frequent customer'

In [22]:
# Adding frequency_flag Regular customer

ords_prods_merge.loc[(ords_prods_merge['frequency_median'] > 10) & (ords_prods_merge['frequency_median'] <= 20), 'frequency_flag'] = 'Regular customer'

In [23]:
# Adding frequency_flag Frequent customer

ords_prods_merge.loc[ords_prods_merge['frequency_median'] <= 10, 'frequency_flag'] = 'Frequent customer'

In [24]:
ords_prods_merge['frequency_flag'].value_counts(dropna=False)

Frequent customer        21560324
Regular customer          7209005
Non-frequent customer     3636707
NaN                             5
Name: frequency_flag, dtype: int64

In [25]:
# Marking 5,127 rows with outliers as missing since  they don’t make sense in terms of the other values in the column

ords_prods_merge.loc[ords_prods_merge['prices'] >100, 'prices'] = np.nan

In [26]:
ords_prods_merge['prices'].max()

25.0

In [31]:
# 09. Export your dataframe as a pickle file and store it correctly in your “Prepared Data” folder.

In [27]:
ords_prods_merge.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_merged_4.8.1.pkl'))

In [28]:
ords_prods_merge.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   3139998      138            28                   6                 11   
1   1977647      138            30                   6                 17   
2    389851      709             2                   0                 21   
3    652770      764             1                   3                 13   
4   1813452      764             3                   4                 17   

   ...    price_range_loc     busiest_day    busiest_days  \
0  ...  Mid-range product  Regularly busy  Regularly busy   
1  ...  Mid-range product  Regularly busy  Regularly busy   
2  ...  Mid-range product     Busiest day    Busiest days   
3  ...  Mid-range product  Regularly busy    Slowest days   
4  ...  Mid-range product      Least busy    Slowest days   

  busiest_period_of_day max_order      loyalty_flag average_price  \
0        Average orders        32  Regular customer      6.935811   
1        Average orders        32  Regular customer      6.935811   
2        Average orders         5      New customer      7.930208   
3        Average orders         3      New customer      4.972414   
4        Average orders         3      New customer      4.972414   

  spenders_flag  frequency_median     frequency_flag  
0   Low spender               8.0  Frequent customer  
1   Low spender               8.0  Frequent customer  
2   Low spender               8.0  Frequent customer  
3   Low spender               9.0  Frequent customer  
4   Low spender               9.0  Frequent customer  

[5 rows x 24 columns]